modules/apps/bookmarks/bookmarks-item-selector-web/src/main/java/com/liferay/bookmarks/item/selector/web/internal/BookmarksItemSelectorView.java

```diff
@@ -0,0 +1,143 @@
+import com.liferay.item.selector.criteria.URLItemSelectorReturnType;
+import com.liferay.portal.kernel.language.Language;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.ListUtil;
+
+import java.io.IOException;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+
+import javax.portlet.PortletURL;
+
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Yang Cao
+ */
+@Component(
+	enabled = false, property = "item.selector.view.order:Integer=100",
+	service = ItemSelectorView.class
+)
+public class BookmarksItemSelectorView
+	implements ItemSelectorView<BookmarksItemSelectorCriterion> {
+
+	@Override
+	public Class<BookmarksItemSelectorCriterion>
+		getItemSelectorCriterionClass() {
+
+		return BookmarksItemSelectorCriterion.class;
+	}
+
+	public ServletContext getServletContext() {
+		return _servletContext;
+	}
+
+	@Override
+	public List<ItemSelectorReturnType> getSupportedItemSelectorReturnTypes() {
+		return _supportedItemSelectorReturnTypes;
+	}
+
+	@Override
+	public String getTitle(Locale locale) {
+		return _language.get(locale, "bookmark-images");
+	}
+
+	@Override
+	public boolean isVisible(ThemeDisplay themeDisplay) {
+		return true;
+	}
+
+	@Override
+	public void renderHTML(
+			ServletRequest servletRequest, ServletResponse servletResponse,
+			BookmarksItemSelectorCriterion bookmarksItemSelectorCriterion,
+			PortletURL portletURL, String itemSelectedEventName, boolean search)
+		throws IOException, ServletException {
+
+		BookmarksItemSelectorViewDisplayContext
+			bookmarksItemSelectorViewDisplayContext =
+				new BookmarksItemSelectorViewDisplayContext(
+					bookmarksItemSelectorCriterion, this,
+					_itemSelectorReturnTypeResolverHandler,
+					itemSelectedEventName, search, portletURL,
+					_bookmarksEntryLocalService);
+
+		servletRequest.setAttribute(
+			BookmarksItemSelectorViewDisplayContext.class.getName(),
+			bookmarksItemSelectorViewDisplayContext);
+
+		ServletContext servletContext = getServletContext();
+
+		RequestDispatcher requestDispatcher =
+			servletContext.getRequestDispatcher("/bookmark_images.jsp");
+
+		requestDispatcher.include(servletRequest, servletResponse);
+	}
+
+	@Reference(unbind = "-")
+	public void setItemSelectorReturnTypeResolverHandler(
+		ItemSelectorReturnTypeResolverHandler
+			itemSelectorReturnTypeResolverHandler) {
+
+		_itemSelectorReturnTypeResolverHandler =
+			itemSelectorReturnTypeResolverHandler;
+	}
+
+	@Reference(
+		target = "(osgi.web.symbolicname=com.liferay.bookmarks.item.selector.web)",
+		unbind = "-"
+	)
+	public void setServletContext(ServletContext servletContext) {
+		_servletContext = servletContext;
+	}
```

Comment:
Avoid mixing method and field injection unless necessary. Pick one and stick to it (in the same class).

Reply:
all updated, could you review these again? thank you!